### ISSUE TITLE
SSL verification error at depth 0: unable to get local issuer certificate (20)

### ISSUE THREAD

**Issue by GamertodArk** [rubygems/rubygems#2231]
I don't know anything about ruby and gems, the reason that installed it it's because it is necesary to install sass, but when i run this command

```
gem install sass
```

This shows up 

```
ERROR:  SSL verification error at depth 0: unable to get local issuer certificate (20)
ERROR:  You must add /C=EN/CN=e9be946d3338ec7a 2 to your local trusted store
ERROR:  SSL verification error at depth 0: unable to get local issuer certificate (20)
ERROR:  You must add /C=EN/CN=e9be946d3338ec7a 2 to your local trusted store
ERROR:  SSL verification error at depth 0: unable to get local issuer certificate (20)
ERROR:  You must add /C=EN/CN=e9be946d3338ec7a 2 to your local trusted store
ERROR:  Could not find a valid gem 'sass' (>= 0), here is why:
          Unable to download data from https://rubygems.org/ - SSL_connect returned=1 errno=0 state=error: certificate verify failed (unable to get local issuer certificate) (https://api.rubygems.org/specs.4.8.gz)
ERROR:  SSL verification error at depth 0: unable to get local issuer certificate (20)
ERROR:  You must add /C=EN/CN=e9be946d3338ec7a 2 to your local trusted store
ERROR:  SSL verification error at depth 0: unable to get local issuer certificate (20)
ERROR:  You must add /C=EN/CN=e9be946d3338ec7a 2 to your local trusted store
```
i have already tried this tutorial
```
http://guides.rubygems.org/ssl-certificate-update/#manual-solution-to-ssl-issue
```

but i didn't help, what's going on?

My environment settings are:
```
C:\> ruby -v
ruby 2.5.0p0 (2017-12-25 revision 61468) [i386-mingw32]

C:\> gem -v
2.7.3

Windows 7 32 bits
```

**Comment by indirect** [comment]
It sounds like your Ruby install might not be configured to know where your SSL certificates are. Can you try following the steps in this guide and report back with the results? http://bundler.io/v1.16/guides/rubygems_tls_ssl_troubleshooting_guide.html

**Comment by GamertodArk** [comment]
Ok, let's see.

The guide said that needed to run this command 
```
 curl -sL https://git.io/vQhWq | ruby
```
to know whether my errors result from the SSL certs issue or the TLS versions issue. But i did't show anything, I run this command twice but in neither of them showed something. Just the prompt shows to write another command again.

I tried to update the bundler running this command
```
 gem install bundler
```

But the same error showed up
```
ERROR:  SSL verification error at depth 0: unable to get local issuer certificate (20)
ERROR:  You must add /C=EN/CN=e9be946d3338ec7a 2 to your local trusted store
ERROR:  SSL verification error at depth 0: unable to get local issuer certificate (20)
ERROR:  You must add /C=EN/CN=e9be946d3338ec7a 2 to your local trusted store
ERROR:  SSL verification error at depth 0: unable to get local issuer certificate (20)
ERROR:  You must add /C=EN/CN=e9be946d3338ec7a 2 to your local trusted store
ERROR:  Could not find a valid gem 'bundler' (>= 0), here is why:
          Unable to download data from https://rubygems.org/ - SSL_connect returned=1 errno=0 state=error: certificate verify failed (unable to get local issuer certificate        ) (https://api.rubygems.org/specs.4.8.gz)
ERROR:  SSL verification error at depth 0: unable to get local issuer certificate (20)
ERROR:  You must add /C=EN/CN=e9be946d3338ec7a 2 to your local trusted store
ERROR:  SSL verification error at depth 0: unable to get local issuer certificate (20)
ERROR:  You must add /C=EN/CN=e9be946d3338ec7a 2 to your local trusted store
```

Then I tried to update RubyGems. I downloaded the lastest ```.gem``` update that i could find, it was the ```2.7.6```, before that version i was using the ```2.7.3```, i don't know it would do a huge change, but i did it just because. Ok i updated to that version and then i checked it with ```gem -v ``` and the right version was installed, but when i run the command to install sass the same error shows up again

```
ERROR:  SSL verification error at depth 0: unable to get local issuer certificate (20)
ERROR:  You must add /C=EN/CN=e9be946d3338ec7a 2 to your local trusted store
ERROR:  SSL verification error at depth 0: unable to get local issuer certificate (20)
ERROR:  You must add /C=EN/CN=e9be946d3338ec7a 2 to your local trusted store
ERROR:  SSL verification error at depth 0: unable to get local issuer certificate (20)
ERROR:  You must add /C=EN/CN=e9be946d3338ec7a 2 to your local trusted store
ERROR:  Could not find a valid gem 'sass' (>= 0), here is why:
          Unable to download data from https://rubygems.org/ - SSL_connect returned=1 errno=0 state=error: certificate verify failed (unable to get local issuer certificate) (https://api.rubygems.org/specs.4.8.gz)
ERROR:  SSL verification error at depth 0: unable to get local issuer certificate (20)
ERROR:  You must add /C=EN/CN=e9be946d3338ec7a 2 to your local trusted store
ERROR:  SSL verification error at depth 0: unable to get local issuer certificate (20)
ERROR:  You must add /C=EN/CN=e9be946d3338ec7a 2 to your local trusted store
```
Then the guide said to install new CA certificates, I did it (again) but i didn't help. 

I don't know what's hapenning, i tried to reintall ruby with the ruby installer like three times... God

```
$ gem env
RubyGems Environment:
  - RUBYGEMS VERSION: 2.7.6
  - RUBY VERSION: 2.5.0 (2017-12-25 patchlevel 0) [i386-mingw32]
  - INSTALLATION DIRECTORY: E:/Ruby25/lib/ruby/gems/2.5.0
  - USER INSTALLATION DIRECTORY: E:/Users/Gamertod/.gem/ruby/2.5.0
  - RUBY EXECUTABLE: E:/Ruby25/bin/ruby.exe
  - EXECUTABLE DIRECTORY: E:/Ruby25/bin
  - SPEC CACHE DIRECTORY: E:/Users/Gamertod/.gem/specs
  - SYSTEM CONFIGURATION DIRECTORY: E:/ProgramData
  - RUBYGEMS PLATFORMS:
    - ruby
    - x86-mingw32
  - GEM PATHS:
     - E:/Ruby25/lib/ruby/gems/2.5.0
     - E:/Users/Gamertod/.gem/ruby/2.5.0
  - GEM CONFIGURATION:
     - :update_sources => true
     - :verbose => true
     - :backtrace => false
     - :bulk_threshold => 1000
  - REMOTE SOURCES:
     - https://rubygems.org/
  - SHELL PATH:
     - E:\Users\Gamertod\bin
     - E:\Program Files\Git\mingw32\bin
     - E:\Program Files\Git\usr\local\bin
     - E:\Program Files\Git\usr\bin
     - E:\Program Files\Git\usr\bin
     - E:\Program Files\Git\mingw32\bin
     - E:\Program Files\Git\usr\bin
     - E:\Users\Gamertod\bin
     - E:\Windows\system32
     - E:\Windows
     - E:\Windows\System32\Wbem
     - E:\Windows\System32\WindowsPowerShell\v1.0
     - E:\Program Files\Git\cmd
     - E:\Program Files\nodejs
     - E:\Ruby25\bin
     - C:\Ruby25\bin
     - E:\Users\Gamertod\AppData\Roaming\npm
     - E:\Program Files\MongoDB\Server\3.2\bin
     - E:\php7
     - E:\Program Files\MySQL\MySQL Server 5.7\bin
     - E:\Program Files\Git\usr\bin\vendor_perl
     - E:\Program Files\Git\usr\bin\core_perl

```

**Comment by GamertodArk** [comment]
Honestly, I think that it isn't a RubyGems's issue, because i have MesaSync Installed and sometimes an error shows up saying something about the SSL certificates like this error, but i don't know why, and even is it isn't a RubyGems error I still need help, so please help me

**Comment by bronzdoc** [comment]
Sorry @GamertodArk this issue tracker is for Rubygems, if your problem is not related to Rubygems i must close this issue, we can't help you with MegaSync  beacuse we don't know nothing about it.

**Comment by indirect** [comment]
@GamertodArk I am not sure if this will help you, but the error message definitely makes it seem like Ruby cannot find the CA certificates, even though you have installed them. You might be able to get more help troubleshooting by going to superuser.com and posting under the "windows" tag. Good luck figuring it out!

**Comment by SDP190** [comment]
I have experienced the same issue on Windows 10 os (I am using a Web-Filter which proxies web requests via a Certificate **Livigent CA**):
```
C:\Users\SDP>gem install rake
ERROR:  SSL verification error at depth 0: unable to get local issuer certificate (20)
ERROR:  You must add /C=US/CN=Livigent CA to your local trusted store
ERROR:  SSL verification error at depth 0: unable to get local issuer certificate (20)
ERROR:  You must add /C=US/CN=Livigent CA to your local trusted store
ERROR:  SSL verification error at depth 0: unable to get local issuer certificate (20)
ERROR:  You must add /C=US/CN=Livigent CA to your local trusted store
ERROR:  Could not find a valid gem 'rake' (>= 0), here is why:
          Unable to download data from https://rubygems.org/ - SSL_connect returned=1 errno=0 state=error: certificate verify failed (unable to get local issuer certificate) (https://rubygems.org/specs.4.8.gz)
ERROR:  SSL verification error at depth 0: unable to get local issuer certificate (20)
ERROR:  You must add /C=US/CN=Livigent CA to your local trusted store
ERROR:  SSL verification error at depth 0: unable to get local issuer certificate (20)
ERROR:  You must add /C=US/CN=Livigent CA to your local trusted store
```
, could it possibly be related to a bad/unrecognized certificate?

**Comment by MSP-Greg** [comment]
@SDP190

Have you tried running:
https://raw.githubusercontent.com/rubygems/ruby-ssl-check/master/check.rb

Or, can you connect with Ruby Net::HTTP using ssl?

**Comment by khiav223577** [comment]
I had this issue before, and solved it by upgrading `rubygems-update` from 2.6.12 to 2.7.11.

**Comment by ashish91** [comment]
I was facing the issue due to `.gemrc`:

```
---
:update_sources: true
:sources:
- https://gems.rubyforge.org/
- https://gems.github.com
:benchmark: false
:bulk_threshold: 1000
:backtrace: false
:verbose: true
gem: --no-ri --no-rdoc
```
